cpp/src/stream_compaction/filter/filter.cu

```diff
@@ -0,0 +1,351 @@
+  }
+};
+
+/// @brief A gather-filter operation that filters the elements of an input range base on a
+/// pre-computed stencil. The stencil serves as both a stencil and a gather map.
+template <typename InputIterator>
+auto filter_by(InputIterator input_begin,
+               InputIterator input_end,
+               cudf::size_type num_selected,
+               cudf::size_type const* stencil,
+               rmm::cuda_stream_view stream,
+               rmm::device_async_resource_ref mr)
+{
+  rmm::device_uvector<typename std::iterator_traits<InputIterator>::value_type> output(
+    static_cast<size_t>(num_selected), stream, mr);
+
+  auto output_size = cuda::std::distance(output.begin(),
+                                         thrust::copy_if(rmm::exec_policy(stream),
+                                                         input_begin,
+                                                         input_end,
+                                                         stencil,
+                                                         output.begin(),
+                                                         filter_predicate{}));
+
+  CUDF_EXPECTS(output_size == num_selected,
+               "The number of selected items does not match the expected count.",  // < This should
+                                                                                   // never happen
+               std::runtime_error);
+
+  return output;
+}
+
+template <bool result_has_nulls>
+struct filter_dispatcher;
+
+template <>
+struct filter_dispatcher<false> {
+  template <typename T>
+    requires(cudf::is_rep_layout_compatible<T>() && cudf::is_fixed_width<T>())
+  std::unique_ptr<cudf::column> operator()(cudf::column_device_view const& col,
+                                           cudf::size_type num_selected,
+                                           cudf::size_type const* stencil_iterator,
+                                           rmm::cuda_stream_view stream,
+                                           rmm::device_async_resource_ref mr) const
+  {
+    auto filtered = filter_by(
+      col.data<T>(), col.data<T>() + col.size(), num_selected, stencil_iterator, stream, mr);
+    auto filtered_size = filtered.size();
+    auto out =
+      cudf::column(col.type(), filtered_size, filtered.release(), rmm::device_buffer{}, 0, {});
+    return std::make_unique<cudf::column>(std::move(out));
+  }
+
+  template <typename T>
+    requires(std::is_same_v<T, cudf::string_view>)
+  std::unique_ptr<cudf::column> operator()(cudf::column_device_view const& col,
+                                           cudf::size_type num_selected,
+                                           cudf::size_type const* stencil_iterator,
+                                           rmm::cuda_stream_view stream,
+                                           rmm::device_async_resource_ref mr) const
+  {
+    auto filtered = filter_by(col.begin<cudf::string_view>(),
```

Comment:
Instead of creating an intermediate string_view vector, we should be able to materialize directly into a column